ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala

```diff
@@ -233,6 +237,20 @@ private[backend] object StorageBackendTestValues {
       rejection_status_code = None,
       rejection_status_message = None,
       rejection_status_details = None,
+      submission_id = submissionId,
+      deduplication_offset = deduplication.flatMap {
+        case Left3(offset) => Some(offset)
+        case _ => None
+      },
+      deduplication_time_seconds = deduplication.flatMap {
+        case Middle3((seconds, _)) => Some(seconds)
+        case _ => None
+      },
+      deduplication_time_nanos = deduplication.flatMap {
+        case Middle3((_, nanos)) => Some(nanos)
+        case _ => None
+      },
```

Comment:
Could extract these together and bind-deconstruct them earlier, maybe slightly better.

Reply:
You could maybe do something like:

```scala
      deduplication_offset = deduplication.leftOr(None)(Some),
      deduplication_time_seconds = deduplication.middleOr(None)(Some),
      deduplication_time_nanos = deduplication.rightOr(None)(Some),
```

That said, I think I'd prefer a custom set of classes, with appropriate getters.

Reply:
It would become even more cryptic:

```
      deduplication_offset = deduplication.flatMap(_.leftOr(None[_])(Some(_))),
      deduplication_time_seconds = deduplication.flatMap(_.middleOr(None[_])(Some(_))).map(_._1),
      deduplication_time_nanos = deduplication.flatMap(_.middleOr(None[_])(Some(_))).map(_._2),
```

I think I'll go with an ADT.